File: integration_tests/typescript_client_tests_real/test/custom-queries.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { GraphLinkClient } from '../lib/generated/client/graph-link-client.js';
import { UserStatus } from '../lib/generated/enums/user-status.js';
import type { UserResult } from '../lib/generated/types/user-result.js';
import type { PostResult } from '../lib/generated/types/post-result.js';
import { newClient } from './real-server-adapter.ts';

let client: GraphLinkClient;
beforeEach(() => { client = newClient(); });

// ── fetchUserAndPost ──────────────────────────────────────────────────────────

describe('fetchUserAndPost', () => {
  it('res.user.name is correct', async () => {
    const res = await client.queries.fetchUserAndPost({ userId: 'user-1', postId: 'post-1' });
    expect(res.user.name).toBe('Alice Smith');
  });

  it('res.user.id is correct', async () => {
    const res = await client.queries.fetchUserAndPost({ userId: 'user-1', postId: 'post-1' });
    expect(res.user.id).toBe('user-1');
  });

  it('res.post.title is correct', async () => {
    const res = await client.queries.fetchUserAndPost({ userId: 'user-1', postId: 'post-1' });
    expect(res.post.title).toBe('Hello World');
  });

  it('res.post.author.id is correct', async () => {
    const res = await client.queries.fetchUserAndPost({ userId: 'user-1', postId: 'post-1' });
    expect(res.post.author.id).toBe('user-1');
  });
});

// ── fetchUserSummary — projected type ─────────────────────────────────────────

describe('fetchUserSummary — projected type', () => {
  it('id is accessible', async () => {
    const res = await client.queries.fetchUserSummary({ id: 'user-1' });
    expect(res.getUser.id).toBe('user-1');
  });

  it('name is accessible', async () => {
    const res = await client.queries.fetchUserSummary({ id: 'user-1' });
    expect(res.getUser.name).toBe('Alice Smith');
  });

  it('status is deserialized as enum', async () => {
    const res = await client.queries.fetchUserSummary({ id: 'user-1' });
    expect(res.getUser.status).toBe(UserStatus.ACTIVE);
  });
});

// ── fetchCachedPair — data correctness ────────────────────────────────────────

describe('fetchCachedPair — data correctness', () => {
  it('user and post data is correct', async () => {
    const res = await client.queries.fetchCachedPair({ userId: 'user-1', postId: 'post-1' });
    expect(res.user.name).toBe('Alice Smith');
    expect(res.post.title).toBe('Hello World');
  });

  it('second call with same args returns correct data', async () => {
    await client.queries.fetchCachedPair({ userId: 'user-1', postId: 'post-1' });
    const res = await client.queries.fetchCachedPair({ userId: 'user-1', postId: 'post-1' });
    expect(res.user.name).toBe('Alice Smith');
    expect(res.post.title).toBe('Hello World');
  });
});

// ── runSearch — custom query, union dispatch ──────────────────────────────────
// Note: the auto-generated search query uses a fragment structure that Spring
// rejects; use runSearch (custom operation) for union dispatch testing.

describe('runSearch — __typename dispatch', () => {
  it('search list has two elements', async () => {
    const res = await client.queries.runSearch({ term: 'test' });
    expect(res.search).toHaveLength(2);
  });

  it('first element has UserResult fields', async () => {
    const res = await client.queries.runSearch({ term: 'test' });
    expect('name' in res.search[0]).toBe(true);
  });

  it('second element has PostResult fields', async () => {
    const res = await client.queries.runSearch({ term: 'test' });
    expect('title' in res.search[1]).toBe(true);
  });

  it('UserResult has correct name', async () => {
    const res = await client.queries.runSearch({ term: 'test' });
    const userResult = res.search[0] as UserResult;
    expect(userResult.name).toBe('Alice Smith');
  });

  it('UserResult has correct email', async () => {
    const res = await client.queries.runSearch({ term: 'test' });
    const userResult = res.search[0] as UserResult;
    expect(userResult.email).toBe('alice@test.com');
  });

  it('PostResult has correct title', async () => {
    const res = await client.queries.runSearch({ term: 'test' });
    const postResult = res.search[1] as PostResult;
    expect(postResult.title).toBe('Hello World');
  });
});

File: integration_tests/dart_client_tests_real/test/custom_queries_test.dart
import 'package:test/test.dart';
import 'package:graphlink_client_integration_tests/generated/client/graph_link_client.dart';
import 'package:graphlink_client_integration_tests/generated/types/user_result.dart';
import 'package:graphlink_client_integration_tests/generated/types/post_result.dart';
import 'package:graphlink_client_integration_tests/generated/enums/user_status.dart';
import 'real_server_adapter.dart';

void main() {
  late GraphLinkClient client;
  setUp(() => client = newClient());

  // ── fetchUserAndPost ───────────────────────────────────────────────────────

  group('fetchUserAndPost', () {
    test('res.user.name is Alice Smith', () async {
      final res = await client.queries
          .fetchUserAndPost(userId: 'user-1', postId: 'post-1', limit: 10);
      expect(res.user.name, equals('Alice Smith'));
    });

    test('res.user.id is user-1', () async {
      final res = await client.queries
          .fetchUserAndPost(userId: 'user-1', postId: 'post-1', limit: 10);
      expect(res.user.id, equals('user-1'));
    });

    test('res.post.title is Hello World', () async {
      final res = await client.queries
          .fetchUserAndPost(userId: 'user-1', postId: 'post-1', limit: 10);
      expect(res.post.title, equals('Hello World'));
    });

    test('res.post.author.id is user-1', () async {
      final res = await client.queries
          .fetchUserAndPost(userId: 'user-1', postId: 'post-1', limit: 10);
      expect(res.post.author.id, equals('user-1'));
    });
  });

  // ── fetchUserSummary — projected type ─────────────────────────────────────

  group('fetchUserSummary — projected type', () {
    test('id is accessible', () async {
      final res = await client.queries.fetchUserSummary(id: 'user-1');
      expect(res.getUser.id, equals('user-1'));
    });

    test('name is accessible', () async {
      final res = await client.queries.fetchUserSummary(id: 'user-1');
      expect(res.getUser.name, equals('Alice Smith'));
    });

    test('status is deserialized as enum', () async {
      final res = await client.queries.fetchUserSummary(id: 'user-1');
      expect(res.getUser.status, equals(UserStatus.ACTIVE));
    });
  });

  // ── fetchCachedPair ────────────────────────────────────────────────────────

  group('fetchCachedPair — data correctness', () {
    test('user and post data is correct', () async {
      final res = await client.queries
          .fetchCachedPair(userId: 'user-1', postId: 'post-1', limit: 10);
      expect(res.user.name, equals('Alice Smith'));
      expect(res.post.title, equals('Hello World'));
    });

    test('second call with same args returns correct data', () async {
      await client.queries
          .fetchCachedPair(userId: 'user-1', postId: 'post-1', limit: 10);
      final res = await client.queries
          .fetchCachedPair(userId: 'user-1', postId: 'post-1', limit: 10);
      expect(res.user.name, equals('Alice Smith'));
      expect(res.post.title, equals('Hello World'));
    });
  });

  // ── runSearch — custom query, union dispatch ──────────────────────────────
  // Note: the auto-generated search query uses a fragment structure that Spring
  // rejects; use runSearch (custom operation) for union dispatch testing.

  group('runSearch — __typename dispatch', () {
    test('search list has two elements', () async {
      final res = await client.queries.runSearch(term: 'test');
      expect(res.search, hasLength(2));
    });

    test('first element is dispatched as UserResult', () async {
      final res = await client.queries.runSearch(term: 'test');
      expect(res.search.first, isA<UserResult>());
    });

    test('second element is dispatched as PostResult', () async {
      final res = await client.queries.runSearch(term: 'test');
      expect(res.search.last, isA<PostResult>());
    });

    test('UserResult has correct name', () async {
      final res = await client.queries.runSearch(term: 'test');
      final userResult = res.search.first as UserResult;
      expect(userResult.name, equals('Alice Smith'));
    });

    test('PostResult has correct title', () async {
      final res = await client.queries.runSearch(term: 'test');
      final postResult = res.search.last as PostResult;
      expect(postResult.title, equals('Hello World'));
    });
  });
}
